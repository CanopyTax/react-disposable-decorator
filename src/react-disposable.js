import React from "react";
import ReactDOM from "react-dom";
import { Observable } from "rx";

const inTestingEnv = typeof jasmine !== "undefined";

export function DisposableDecorator() {
	return function decorator(DecoratedComponent) {
		return class Disposable extends React.Component {
			disposables = [];
			render() {
				return (
					<DecoratedComponent
						{...this.props}
						cancelWhenUnmounted={this.cancelWhenUnmounted}
						cancelAllSubscriptions={this.cancelAllSubscriptions}
					/>
				);
			}
			componentWillUnmount() {
				this.cancelAllSubscriptions();
			}
			cancelWhenUnmounted = cancelWhenUnmounted.bind(this);
			cancelAllSubscriptions = cancelAllSubscriptions.bind(this);
		};
	};
}

export function disposableModal(El, props = {}) {
	let el;

	return Observable.create(observer => {
		el = document.createElement("div");
		document.body.appendChild(el);

		ReactDOM.render(
			<El {...props} close={observer.onCompleted} save={observer.onNext} />,
			el
		);
	}).finally(() => {
		ReactDOM.unmountComponentAtNode(el);
		el.parentNode.removeChild(el);
	});
}

export class DisposableElement extends React.Component {
	disposables = [];
	render() {
		return this.props.children(
			this.cancelWhenUnmounted,
			this.cancelAllSubscriptions
		);
	}
	componentWillUnmount() {
		this.cancelAllSubscriptions();
	}
	cancelWhenUnmounted = cancelWhenUnmounted.bind(this);
	cancelAllSubscriptions = cancelAllSubscriptions.bind(this);
}

function cancelWhenUnmounted(...thingsToCancel) {
	thingsToCancel.forEach(thingToCancel => {
		if (!thingToCancel || typeof thingToCancel.dispose !== "function") {
			throw new Error(
				`cancelWhenUnmounted should be called with one or more disposables (an object with a dispose function)`
			);
		}
		this.disposables.push(thingToCancel);
	});
}

function cancelAllSubscriptions() {
	this.disposables.forEach(disposable => {
		if (disposable && typeof disposable.dispose === "function") {
			disposable.dispose();
		}
	});

	this.disposables = [];
}
