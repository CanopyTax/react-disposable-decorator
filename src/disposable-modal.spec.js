import { DisposableDecorator, disposableModal } from "./react-disposable.js";
import { Observable } from "rx";
import { mount } from "enzyme";
import React from "react";

describe("DisposableElement", function() {
	it("should render", function() {
		function Component({ cancelWhenUnmounted, cancelAllSubscriptions, stuff }) {
			return <div>Hello {stuff}</div>;
		}

		const disposable = disposableModal(Component).subscribe(data => {
		});

		expect(document.body).toMatchSnapshot();
		disposable.dispose();
	});

	it("should provide close method", function(done) {
		class Component extends React.Component {
			componentDidMount() {
				setTimeout(this.props.close, 100);
			}
			render() {
				return <div>Hello {this.props.stuff}</div>;
			}
		}

		const disposable = disposableModal(Component).subscribe(data => {});

		setTimeout(() => {
			expect(document.body).toMatchSnapshot();
			disposable.dispose();
			done();
		}, 200);
	});

	it('should work with disposable decorator', function() {
		class Component extends React.Component {
			componentDidMount() {
				this.props.cancelWhenUnmounted(
					disposableModal(Modal).subscribe(data => {})
				);
			}
			render() {
				return <div>Hello {this.props.stuff}</div>;
			}
		}

		class Modal extends React.Component {
			render() {
				return <button close={this.props.close}></button>
			}
		}

		const Decorated = DisposableDecorator()(Component);
		const wrapper = mount(<Decorated stuff={1} />);
		expect(document.body).toMatchSnapshot();
		wrapper.unmount();
		expect(document.body).toMatchSnapshot();
	});
});
