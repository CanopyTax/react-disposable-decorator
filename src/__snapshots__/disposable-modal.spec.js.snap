// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisposableElement should provide close method 1`] = `
<body>
  <div>
    <div>
      Hello 
    </div>
  </div>
</body>
`;

exports[`DisposableElement should render 1`] = `
<body>
  <div>
    <div>
      Hello 
    </div>
  </div>
</body>
`;

exports[`DisposableElement should work with disposable decorator 1`] = `
<body>
  <div>
    <button />
  </div>
</body>
`;

exports[`DisposableElement should work with disposable decorator 2`] = `<body />`;
